vllm/core/scheduler.py

```diff
@@ -1584,64 +1636,174 @@ def _get_num_lookahead_slots(self, is_prefill: bool,
+
+        return num_uncached_new_tokens_seq, num_cached_new_tokens_seq
+
+    def _chunk_new_tokens_to_schedule(
+        self,
+        budget: SchedulingBudget,
+        prompt_limit: int,
+        num_new_tokens: int,
+    ) -> int:
+        """
+        Chunks the number of new tokens to schedule based on the budget when
+        chunked prefill is enabled.
+
+        Args:
+            budget: The budget to chunk the number of tokens to compute.
+            prompt_limit: The maximum number of tokens allowed in a prompt.
+            num_new_tokens: The number of new tokens to schedule.
+
+        Returns:
+            The number of new tokens to schedule after chunking.
+        """
+        remaining_token_budget = budget.remaining_token_budget()
+        if self.scheduler_config.is_multi_step:
+            # The current multi-step + chunked prefill capability does
+            # not actually support chunking prompts.
+            #
+            # Therefore, `num_new_tokens` is computed in the same fashion
+            # for both multi-step+chunked-prefill &
+            # multi-step+chunked-prefill+APC
+            #
+            # Prompts with more tokens than the current remaining budget
+            # are postponed to future scheduler steps
+            if num_new_tokens > prompt_limit:
+                # If the seq_group is in prompt-stage, pass the
+                # num_new_tokens as-is so the caller can ignore
+                # the sequence.
+                pass
             else:
-                num_new_tokens = min(num_new_tokens, remaining_token_budget)
+                num_new_tokens = (0 if num_new_tokens > remaining_token_budget
+                                  else num_new_tokens)
+        elif self.cache_config.enable_prefix_caching:
+            # When prefix caching is enabled, we always allocate
+            # the number of new tokens that is dividable by the block
+            # size to avoid partial block matching.
+            block_size = self.cache_config.block_size
+            remainder = budget.token_budget % block_size
+            if remainder != 0:
+                raise ValueError("When enabling chunked prefill and "
+                                 "prefix caching, max_num_batched_tokens "
+                                 "(chunk size) must be dividable by "
+                                 "block size, but got chunk_size "
+                                 f"({budget.token_budget}) % block_size "
+                                 f"({block_size}) = {remainder}")
+            # Round down to block
+            remaining_token_budget = (remaining_token_budget // block_size *
+                                      block_size)
+            # Calculate the number of new tokens that are not cached with
+            # chunk cap.
+            num_new_tokens = min(num_new_tokens, remaining_token_budget)
+        else:
+            num_new_tokens = min(num_new_tokens, remaining_token_budget)
```

Comment:
nit: Avoid repeating the last line, but I don't have strong preference.

```suggestion
        else:
            if self.cache_config.enable_prefix_caching:
                # When prefix caching is enabled, we always allocate
                # the number of new tokens that is dividable by the block
                # size to avoid partial block matching.
                block_size = self.cache_config.block_size
                remainder = budget.token_budget % block_size
                if remainder != 0:
                    raise ValueError("When enabling chunked prefill and "
                                     "prefix caching, max_num_batched_tokens "
                                     "(chunk size) must be dividable by "
                                     "block size, but got chunk_size "
                                     f"({budget.token_budget}) % block_size "
                                     f"({block_size}) = {remainder}")
                # Round down to block
                remaining_token_budget = (remaining_token_budget // block_size *
                                          block_size)
                # Calculate the number of new tokens that are not cached with
                # chunk cap.
            num_new_tokens = min(num_new_tokens, remaining_token_budget)
```